tikv_metrics_proxy/main.go

```diff
@@ -0,0 +1,228 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"os/signal"
+	"sort"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/kvproto/pkg/debugpb"
+	"github.com/pingcap/tidb-inspect-tools/pkg/utils"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	dto "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/expfmt"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+const (
+	getMetricsTimeout = time.Duration(15) * time.Second
+)
+
+// Exporter exposes Prometheus metrics of TiKV server
+type Exporter struct {
+	client *rpcClient
+	stores []string
+}
+
+type tikvOpts struct {
+	addrs    string
+	security Security
+}
+
+func checkFlags(opts tikvOpts) {
+	if opts.addrs == "" {
+		log.Fatal("missing startup flag: --tikv.addrs")
+	}
+}
+
+// NewExporter returns an initialized Exporter.
+func NewExporter(opts tikvOpts) (*Exporter, error) {
+	stores, err := utils.ParseHostPortAddr(opts.addrs)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	client := newRPCClient(opts.security)
+	for _, store := range stores {
+		_, err := client.getConn(store)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	return &Exporter{client: client, stores: stores}, nil
+}
+
+func sanitizeLabels(
+	metricFamilies map[string]*dto.MetricFamily,
+	groupingLabels map[string]string,
+) {
+	for _, mf := range metricFamilies {
+		for _, m := range mf.GetMetric() {
+			for key, value := range groupingLabels {
+				l := &dto.LabelPair{
+					Name:  proto.String(key),
+					Value: proto.String(value),
+				}
+				m.Label = append(m.Label, l)
+			}
+			sort.Sort(prometheus.LabelPairSorter(m.Label))
+		}
+	}
+}
+
+func (e *Exporter) getMetricFamilies() []*dto.MetricFamily {
+	wg := sync.WaitGroup{}
+	mutex := &sync.Mutex{}
+	allMetrics := make([]*dto.MetricFamily, 0, 1024)
+
+	getStoreMetrics := func(store string) {
+		defer wg.Done()
+
+		tikvConn, err := e.client.getConn(store)
+		if err != nil {
+			return
+		}
+
+		tikvClient := debugpb.NewDebugClient(tikvConn)
+		ctx, cancel := context.WithTimeout(context.Background(), getMetricsTimeout)
+		defer cancel()
+		metrics, err := tikvClient.GetMetrics(ctx, &debugpb.GetMetricsRequest{})
+		if err != nil {
+			log.Errorf("tikv store '%s', get metrics error, %v", store, err)
```

Comment:
do we need to remove the connection if meet any error?

Reply:
grpc client has reconnection mechanism，and DefaultBackoffMaxDelay is 120s, I add `grpc.WithBackoffMaxDelay(time.Second*5)` option in the new commit.